keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,204 @@
+
+- To change or implement additional features in kube-proxy.
+- To deal with graduation of any other component of kube-proxy, other than its configuration.
+- To remove most or even all of the command line flags, that have corresponding component config options.
+
+## Proposal
+
+The idea is to conduct the process of graduation to beta in small steps in the span of at least one Kubernetes release cycle. This will be done by creating one or more alpha versions of the config with the last alpha version being copied as v1beta1 after the community is happy with it.
+Each of the sub-sections below can result in a separate alpha version release, although it will be better for users to have no more than a couple of alpha versions past v1alpha1.
+The proposed process is similar to the already successfully used one for kubeadm.
+
+### Re-encapsulate mode specific options
+
+The current state of the config has proven that:
+- Some options are deemed as mode specific, but are in fact shared between all modes.
+- Some options are placed directly into KubeProxyConfiguration, but are in fact mode specific ones.
+- There are options that are shared between some (but not all) modes. Specific features of the underlying implementation are common and this happens only within the boundaries of the platform (iptables and ipvs modes for example).
+- Although legacy Linux and Windows user mode proxies are separate code bases, they have a common set of options.
+
+With that in mind, the following measures are proposed:
+- Mode specific structs are consolidated to not use fields from other mode specific structs.
+- Introduce a single combined legacy user mode proxy struct for both Linux and Windows backends.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
+
+Platform specific fields need to be grouped together by platform name. This can be done by introducing structs to house them - one per platform (one for Linux, one for Windows, etc.). Then these structs will have an instance in the KubeProxyConfiguration.
+That can allow for users to specify settings for more than one platform per a YAML document, thus allowing for deduplication of platform independent shared fields. This can reduce the possibility of an error for users.
+So far, research shows, that platform dependent settings are also mode dependent too. Hence each platform struct is going to contain a set of pointers to mode structs - those, that are specific to this platform and platform independent modes.
+Selecting a mode of operation is done on a per-platform basis. To select a mode, only one of the mode sub-structures has to be filled in. If no sub-struct is filled in, a default mode of operation is picked along with its default settings.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+platformX: # Only one sub-key can be provided to select a specific mode of operation
```

Comment:
@vllry I think this is what you're talking about right? Just skip the platform indirection and have a mode oneof?